src/cirrus/lib/task.py

```diff
@@ -0,0 +1,235 @@
+
+        # The original items from the payload
+        self.original_items = deepcopy(payload['features'])
+
+        self.items = self._payload['features']
+
+        # set up logger
+        self.logger = get_task_logger(f"task.{self._name}", payload=payload)
+
+        # local mode? 
+        self._local = local
+
+        # create temporary work directory if workdir is None
+        self._workdir = workdir
+        if workdir is None:
+            self._workdir = Path(mkdtemp())
+            self._tmpworkdir = True
+        else:
+            self._workdir = Path(workdir)
+            self._tmpworkdir = False
+            makedirs(self._workdir, exist_ok=True)
+
+    def __del__(self):
+        # remove work directory if not running locally
+        if self._tmpworkdir:
+            self.logger.debug(f"Removing work directory {self._workdir}")
+            rmtree(self._workdir)
+
+    @property
+    def id(self) -> str:
+        return self._payload['id']
+
+    @property
+    def output_payload(self):
+        processing_ext = 'https://stac-extensions.github.io/processing/v1.1.0/schema.json'
+        for i in self.items:
+            i['stac_extensions'].append(processing_ext)
+            i['stac_extensions'] = list(set(i['stac_extensions']))
+            i['properties']['processing:software'] = {
+                self._name: self._version
+            }
+        self._payload['features'] = self.items
+        # add provenance metadata
+
+        return self._payload
+
+    @property
+    def process_definition(self) -> Dict:
+        return self._payload['process']
+
+    @property
+    def parameters(self) -> Dict:
+        return self.process_definition['tasks'].get(self._name, {})
+
+    @property
+    def output_options(self) -> Dict:
+        return self.process_definition.get('output_options', {})
+
+    @classmethod
+    def validate(cls, payload) -> bool:
+        # put validation logic on input Items and process definition here
+        return True
+
+    def download_assets(self, assets: Optional[List[str]]=None):
+        """Download provided asset keys for all items in payload. Assets are saved in workdir in a
+           directory named by the Item ID, and the items are updated with the new asset hrefs.
+
+        Args:
+            assets (Optional[List[str]], optional): List of asset keys to download. Defaults to all assets.
+        """
+        for i, item in enumerate(self.items):
+            outdir = self._workdir / Path(item['id'])
+            makedirs(outdir, exist_ok=True)
+            self.items[i] = download_item_assets(item, path=outdir, assets=assets)
+
+    def upload_assets(self, assets: Optional[List[str]]=None):
+        if self._local:
+            self.logger.warn('Running in local mode, assets not uploaded')
+            return
+        for i, item in enumerate(self.items):
+            self.items[i] = upload_item_assets(item, **self.output_options)
+
+    # this should be in PySTAC
+    @classmethod
```

Comment:
`@staticmethod`? Looks like `self` isn't used.

Reply:
Good call - changed it to staticmethod, but as noted I think this is best implemented in PySTAC as a general way to create new derived items from existing ones.